src/compiler/path.ts

```diff
@@ -624,12 +624,115 @@ export function getNormalizedPathComponents(path: string, currentDirectory: stri
 }
 
 /** @internal */
-export function getNormalizedAbsolutePath(fileName: string, currentDirectory: string | undefined): string {
-    return getPathFromPathComponents(getNormalizedPathComponents(fileName, currentDirectory));
+export function getNormalizedAbsolutePath(path: string, currentDirectory: string | undefined): string {
+    let rootLength = getRootLength(path);
+    if (rootLength === 0 && currentDirectory) {
+        path = combinePaths(currentDirectory, path);
+        rootLength = getRootLength(path);
+    }
+    const simple = simpleNormalizePath(path);
+    if (simple !== undefined) {
+        return simple;
+    }
+    const root = path.substring(0, rootLength);
+    const normalizedRoot = root && normalizeSlashes(root);
+    // `normalized` is only initialized once `path` is determined to be non-normalized
+    let normalized = normalizedRoot === root ? undefined : normalizedRoot;
+    let index = rootLength;
+    let segmentStart = index;
+    let normalizedUpTo = index;
+    let seenNonDotDotSegment = rootLength !== 0;
+    while (index < path.length) {

```

Comment:
Since `path.length` doesn't change, it may be more efficient to store it in a local and reusing that rather than looking it up each time (both here and elsewhere in the function).